src/act_comm.cpp

```diff
@@ -90,23 +91,13 @@ void do_delete(CHAR_DATA *ch, const char *argument) {
 }
 
 void announce(const char *buf, CHAR_DATA *ch) {
-    Descriptor *d;
-
-    if (descriptor_list == nullptr)
-        return;
-
     if (ch->in_room == nullptr)
         return; /* special case on creation */
 
-    for (d = descriptor_list; d != nullptr; d = d->next) {
-        CHAR_DATA *victim;
-
-        victim = d->person();
-
-        if (d->is_playing() && d->character() != ch && victim && can_see(victim, ch)
-            && !IS_SET(victim->comm, COMM_NOANNOUNCE) && !IS_SET(victim->comm, COMM_QUIET)) {
+    for (auto &d : descriptors().all_who_can_see(ch)) {
+        auto *victim = d.person();
```

Comment:
There's a 'victim' check on 106, do we need that?

Reply:
I guess not given your earlier comment.

Reply:
As best I can tell: no. I put some rationale in the PR itself but:
* `all_who_can_see` only returns playing descriptors that can "see" `ch`
* if a descriptor is playing then they must have at least one character associated with them. This I'm 99.998% sure (I can't find how it could not be true but there's a ton of defensive checks like this).
* `person` always returns the real character associated with a descriptor (if a char is switched then it returns `original` else the `character`). So for a "playing" character it's guaranteed to be non-NULL (as is `character()`, `original()` will be null for unswitched folks).

Might be worth putting those checks into the accessors themselves to maintain the invariant, and `throw` if not (though that would shut the MUD down).